### PR TITLE
One keypair

### LINKED ISSUE #671: Remove pbft specific `keypair.h`
Under `libbyz` there is a pbft specific `keypair.h` that is used only in the `libbyz/pbft` files. This should be removed and the one under `tls` should be used instead.

### PR DESCRIPTION
**Pull request by olgavrou**
Resolves #671 

- Removed `libbyz/keypair.h`
- Removed `pub_key_enc`
- Removed all the hard coded private and public keys that where passed to pbft
- Passing the node private key and the node certificates into pbft via nodestate
- Renaming `pub_key_sig` to `cert`
- Replacing the signature verification via the node public key with signature verification via the node certificate (creating a tls::Verifier)
- Updated the standalone pbft e2e test configuration (priv key/cert) to correspond to the `tls/keypair.h`
- Removed `epbft_drng.h` file

Opening as a draft PR as the signature verification is still failing and I am trying to find out why.

### CODE REVIEW

File: cmake/pbft.cmake [hunk 1]
```diff
@@ -65,6 +65,7 @@ if("sgx" IN_LIST TARGET)
     ${PARSED_ARGS_INCLUDE_DIRS}
     ${EVERCRYPT_INC}
   )
+  use_client_mbedtls(libbyz.enclave)
```
Comment (jumaffre) [inline cmake/pbft.cmake]: This should be `ue_oe_mbedtls()`.

File: src/consensus/pbft/libbyz/New_principal.h [hunk 2]
```diff
@@ -20,11 +20,8 @@ struct New_principal_rep : public Message_rep
   uint32_t ip_len;
   char ip[32];
 
-  uint32_t pubk_sig_len;
-  char pubk_sig[128];
-
-  uint32_t pubk_enc_len;
-  char pubk_enc[128];
+  uint32_t cert_len;
+  char cert[128];
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/New_principal.h]: I'm not sure where this is used but 128 seems very small for a cert?

File: src/consensus/pbft/libbyz/Pre_prepare.cpp [hunk 3]
```diff
@@ -291,12 +291,21 @@ bool Pre_prepare::pre_verify()
...
-               rep().digest.digest_size(),
-               (const char*)get_digest_sig().data()))
+      auto sender_principal =
+        pbft::GlobalState::get_node().get_principal(sender);
+
+      if (
+        !sender_principal->has_certificate_set() &&
+        pbft::GlobalState::get_node().f() == 0)
+      {
+        // we have not configured this node yet
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Pre_prepare.cpp]: I think the comment should be more explicit here? e.g. "Do not verify signature of first pre-prepares since node certificate required for verification is contained in the pre-prepare requests".

File: src/consensus/pbft/libbyz/Status.cpp [hunk 4]
```diff
@@ -70,17 +70,11 @@ bool Status::pre_verify()
       // Received message from unknown sender
       LOG_INFO << "Request from unknown pricipal, id:" << id() << std::endl;
 
-      std::string pubk_sig =
-        "aad14ecb5d7ca8caf5ee68d2762721a3d4fdb09b1ae4a699daf74985193b7d42";
-      std::string pubk_enc =
-        "893d4101c5b225c2bdc8633bb322c0ef9861e0c899014536e11196808ffc0d17";
-
       PrincipalInfo info;
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Status.cpp]: Should this be removed altogether as principals are added through kv hooks?

File: src/consensus/pbft/libbyz/test/client_test.cpp [hunk 5]
```diff
@@ -108,13 +108,22 @@ int main(int argc, char** argv)
   }
 
   GeneralInfo general_info = files::slurp_json(config_file);
-  std::string privk = files::slurp_string(privk_file);
+  PrivateKey privk_j = files::slurp_json(privk_file);
   NodeInfo node_info;
+  tls::KeyPairPtr kp = tls::make_key_pair(privk_j.privk);
+  auto node_cert = kp->self_sign("CN=CCF node", tls::SubjectAltName());
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/test/client_test.cpp]: You shouldn't need `tls::SubjectAltName()` here?

File: cmake/pbft.cmake (hunk 1 above)
Comment (achamayou) [reply cmake/pbft.cmake]: In general, use_client_mbedtls() on a target whose name ends in "enclave" is unlikely to be what we want :)

File: src/consensus/pbft/libbyz/New_principal.h (hunk 2 above)
Comment (achamayou) [reply src/consensus/pbft/libbyz/New_principal.h]: Yes, I think this would ideally be a MAX_CERT_something coming from perhaps mbedtls.

File: cmake/pbft.cmake (hunk 1 above)
Comment (olgavrou) [reply cmake/pbft.cmake]: true dat

File: src/consensus/pbft/libbyz/New_principal.h (hunk 2 above)
Comment (achamayou) [reply src/consensus/pbft/libbyz/New_principal.h]: tls::Cert?

File: src/tls/keypair.h [hunk 6]
```diff
@@ -63,6 +63,8 @@ namespace tls
     "-----BEGIN CERTIFICATE-----\n";
   static constexpr auto PEM_CERTIFICATE_FOOTER = "-----END CERTIFICATE-----\n";
 
+  using PbftSignature = std::array<uint8_t, MBEDTLS_ECDSA_MAX_LEN>;
```
Comment (eddyashton) [inline src/tls/keypair.h]: If we're defining `PbftSignatureSize`, we should use it here too (`= std::array<uint8_t, PbftSignatureSize>;`)

`PbftSignatureSize` is a value, not a type, so it should be `pbft_signature_size` (or maybe `PBFT_SIGNATURE_SIZE`, since it is macro-like).

File: src/node/nodestate.h [hunk 7]
```diff
@@ -1420,7 +1420,7 @@ namespace ccf
     void setup_pbft(const CCFConfig& config)
     {
       setup_n2n_channels();
-
+      std::string cert_string(node_cert.begin(), node_cert.end());
```
Comment (jumaffre) [inline src/node/nodestate.h]: Can we keep the certificate as a `std::vector<uint8_t>` throughout PBFT?

File: src/node/nodestate.h [hunk 8]
```diff
@@ -1431,7 +1431,9 @@ namespace ccf
         rpcsessions,
         *network.tables->get<pbft::RequestsMap>(pbft::Tables::PBFT_REQUESTS),
         *network.tables->get<pbft::PrePreparesMap>(
-          pbft::Tables::PBFT_PRE_PREPARES));
+          pbft::Tables::PBFT_PRE_PREPARES),
+        node_kp->private_key_pem().str(),
```
Comment (jumaffre) [inline src/node/nodestate.h]: Same for the private key (`std::vector<uint8_t>`).

File: src/tls/keypair.h [hunk 9]
```diff
@@ -63,6 +63,8 @@ namespace tls
...
+  static constexpr auto PbftSignatureSize = MBEDTLS_ECDSA_MAX_LEN;
```
Comment (jumaffre) [inline src/tls/keypair.h]: Can this be defined somewhere in PBFT as it is PBFT specific?

File: tests/infra/libbyz/create_config.py [hunk 10]
```diff
@@ -8,7 +8,10 @@ def create_private_config(public_ip, private_key):
     config_dir = "config_private"
     os.makedirs(config_dir, exist_ok=True)
     with open(os.path.join(config_dir, public_ip), "w+") as kf:
-        kf.write(private_key)
+        kf.write("{")
+        pk = f'"privk":"{private_key}"'
+        kf.write(pk)
+        kf.write("}")
```
Comment (achamayou) [inline tests/infra/libbyz/create_config.py]:
```
json.dump({"privk": private_key}, kf)
```
Is shorter and will work even if private_key contains a quote character.

Comment (eddyashton) [inline tests/infra/libbyz/create_config.py]: This should use a library rather than manually formatting:
```
o = {"privk": private_key}
kf.write(json.dumps(o))
```

File: src/consensus/pbft/libbyz/New_principal.h (hunk 2 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/New_principal.h]: Yes thanks, this was done via search/rename and I didn't notice it.

File: src/consensus/pbft/libbyz/Status.cpp (hunk 4 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Status.cpp]: It is still needed in order for backups that are joining to accept the first message from the primary, so that they can then accept the pre-prepares, execute them, and have the kv written to

**Comment by cimetrics** [comment]
one_keypair@[3776 aka 20200110.3](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3776&view=results) vs master ewma over 20 builds from [3663](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3663&view=results) to [3772](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3772&view=results)
images

**Comment by cimetrics** [comment]
one_keypair@[3785 aka 20200113.3](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3785&view=results) vs master ewma over 20 builds from [3676](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3676&view=results) to [3781](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3781&view=results)
images

File: src/node/nodestate.h (hunk 8 above)
Comment (olgavrou) [reply src/node/nodestate.h]: I changed `cert` since it already is a `vector<uint8_t>` so there is not need to convert it into a string, but since the private key is already a string and we can use it to create a `keypair` I think it is fine to pass it in as a string

**Comment by cimetrics** [comment]
one_keypair@[3789 aka 20200113.5](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3789&view=results) vs master ewma over 20 builds from [3676](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3676&view=results) to [3781](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3781&view=results)
images

**Comment by cimetrics** [comment]
one_keypair@[3791 aka 20200113.6](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3791&view=results) vs master ewma over 20 builds from [3676](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3676&view=results) to [3781](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3781&view=results)
images

**Comment by cimetrics** [comment]
one_keypair@[3793 aka 20200113.7](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3793&view=results) vs master ewma over 20 builds from [3676](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3676&view=results) to [3781](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3781&view=results)
images

File: src/consensus/pbft/libbyz/Node.cpp [hunk 11]
```diff
@@ -116,16 +115,17 @@ void Node::add_principal(const PrincipalInfo& principal_info)
   {
     LOG_INFO << "Principal with id: " << principal_info.id
              << " has already been configured" << std::endl;
+    auto principal = it->second;
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Node.cpp]: IIUI, we replace the certificate of an existing principal with the certificate of the new "duplicated" (based on id) principal, but nothing else. It seems a little odd. Why is this required?

File: src/consensus/pbft/libbyz/Pre_prepare.cpp [hunk 12]
```diff
@@ -291,12 +287,23 @@ bool Pre_prepare::pre_verify()
...
-               (const char*)get_digest_sig().data()))
+      auto sender_principal =
+        pbft::GlobalState::get_node().get_principal(sender);
+
+      if (
+        !sender_principal->has_certificate_set() &&
+        pbft::GlobalState::get_node().f() == 0)
+      {
+        // Do not verify signature of first pre-prepare since node certificate
+        // required for verification is contained in the pre-prepar requests
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Pre_prepare.cpp]: "... pre-prepar**e** requests"

File: src/consensus/pbft/libbyz/Node.cpp (hunk 11 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Node.cpp]: The principal object will be setup during the first `Status` message (see `Status::pre_verify()`) but the request will be empty. We will get the certificate for this principal when `add_node` is triggered via the `node`s table hook so we want to set the certificate there.

File: src/consensus/pbft/libbyz/Replica.cpp [hunk 13]
```diff
@@ -1582,13 +1582,9 @@ void Replica::handle(New_principal* m)
   LOG_INFO << "recevied new message to add principal, id:" << m->id()
```
Comment (jumaffre) [inline src/consensus/pbft/libbyz/Replica.cpp]: "received"

File: src/consensus/pbft/libbyz/Checkpoint.cpp [hunk 14]
```diff
@@ -18,23 +18,24 @@ Checkpoint::Checkpoint(Seqno s, Digest& d, bool stable) :
...
-    Checkpoint_tag,
-    sizeof(Checkpoint_rep) + pbft::GlobalState::get_node().sig_size())
+  Message(Checkpoint_tag, sizeof(Checkpoint_rep) + pbft_max_signature_size)
 {
 #endif
   rep().extra = (stable) ? 1 : 0;
   rep().seqno = s;
   rep().digest = d;
   rep().id = pbft::GlobalState::get_node().id();
+#ifdef USE_PKEY_CHECKPOINTS
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/Checkpoint.cpp]: That doesn't seem exposed through CMake?

File: src/consensus/pbft/libbyz/Prepare.h [hunk 15]
```diff
@@ -23,9 +23,16 @@ struct Prepare_rep : public Message_rep
...
 #ifdef SIGN_BATCH
-  KeyPair::Signature batch_digest_signature;
+  size_t digest_sig_size;
+  PbftSignature batch_digest_signature;
+  static constexpr size_t padding_size =
+    ALIGNED_SIZE(pbft_max_signature_size) - pbft_max_signature_size;
+  std::array<uint8_t, padding_size> digest_padding;
+#endif
+
+#ifdef USE_PKEY
```
Comment (achamayou) [inline src/consensus/pbft/libbyz/Prepare.h]: What's USE_PKEY?

File: src/consensus/pbft/libbyz/Checkpoint.cpp (hunk 14 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Checkpoint.cpp]: No these are hard coded in `parameters.h`. We should make a pass at the file and remove the ifdefs that are not needed anymore and are always set, but not as a part of this PR :)

File: src/consensus/pbft/libbyz/Prepare.h (hunk 15 above)
Comment (olgavrou) [reply src/consensus/pbft/libbyz/Prepare.h]: Same as above

Comment (achamayou) [reply src/consensus/pbft/libbyz/Prepare.h]: So there's USE_PKEY_CHECKPOINTS and VIEW_CHANGES (which as far as I can tell are always on, and so in my opinion should be removed, because we never test without), but I can't find anything setting USE_PKEY or explaining what it's for, even though it's gating a fair bit of stuff in Prepare and Pre_Prepare. Can we also remove it?

Comment (olgavrou) [reply src/consensus/pbft/libbyz/Prepare.h]: Sure created ticket for this: https://github.com/microsoft/CCF/issues/691

I want to remove these carefully to make sure nothing gets deleted by mistake

Feel free to move it to `experimental ePBFT` milestone if you think it's better there

**Comment by cimetrics** [comment]
one_keypair@[3796 aka 20200113.9](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3796&view=results) vs master ewma over 20 builds from [3676](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3676&view=results) to [3781](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3781&view=results)
images

File: src/consensus/pbft/libbyz/Checkpoint.cpp (hunk 14 above)
Comment (achamayou) [reply src/consensus/pbft/libbyz/Checkpoint.cpp]: Ok!

File: src/consensus/pbft/libbyz/Prepare.h (hunk 15 above)
Comment (achamayou) [reply src/consensus/pbft/libbyz/Prepare.h]: Thank you! I don't think it's a requirement for `experimental ePBFT1, I think it can be done later, along with #562 and so on.

File: src/consensus/pbft/libbyz/Checkpoint.cpp [hunk 16]
```diff
@@ -51,7 +52,8 @@ void Checkpoint::re_authenticate(Principal* p, bool stable)
   if (rep().extra != 1 && stable)
   {
     rep().extra = 1;
-    pbft::GlobalState::get_node().gen_signature(
+    rep().sig_size = 0;
```
Comment (ashamis) [inline src/consensus/pbft/libbyz/Checkpoint.cpp]: this line is not needed

File: src/consensus/pbft/libbyz/Node.cpp (hunk 11 above)
Comment (ashamis) [inline src/consensus/pbft/libbyz/Node.cpp]: auto &principal

File: src/tls/pem.h [hunk 17]
```diff
@@ -19,7 +19,9 @@ namespace tls
     std::string s;
 
   public:
-    Pem(){};
+    Pem() {}
```
Comment (ashamis) [inline src/tls/pem.h]: Pem() = default;

**Comment by cimetrics** [comment]
one_keypair@[3799 aka 20200114.2](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3799&view=results) vs master ewma over 20 builds from [3690](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3690&view=results) to [3797](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3797&view=results)
images
